processing/src/main/java/org/apache/druid/query/scan/ScanQueryQueryToolChest.java

```diff
@@ -131,4 +180,80 @@ public void cleanup(ScanQueryLimitRowIterator iterFromMake)
       }
     };
   }
+
+  @VisibleForTesting
+  Iterator<ScanResultValue> sortAndLimitScanResultValues(Iterator<ScanResultValue> inputIterator, ScanQuery scanQuery)
+  {
+    Comparator<ScanResultValue> priorityQComparator = new ScanResultValueTimestampComparator(scanQuery);
+
+    // Converting the limit from long to int could theoretically throw an ArithmeticException but this branch
+    // only runs if limit < MAX_LIMIT_FOR_IN_MEMORY_TIME_ORDERING (which should be < Integer.MAX_VALUE)
+    int limit = Math.toIntExact(scanQuery.getLimit());
+    PriorityQueue<ScanResultValue> q = new PriorityQueue<>(limit, priorityQComparator);
+
+    while (inputIterator.hasNext()) {
+      ScanResultValue next = inputIterator.next();
+      List<Object> events = (List<Object>) next.getEvents();
+      for (Object event : events) {
+        // Using an intermediate unbatched ScanResultValue is not that great memory-wise, but the column list
+        // needs to be preserved for queries using the compactedList result format
+        q.offer(new ScanResultValue(null, next.getColumns(), Collections.singletonList(event)));
+        if (q.size() > limit) {
+          q.poll();
+        }
+      }
+    }
+    // Need to convert to a List because Priority Queue's iterator doesn't guarantee that the sorted order
+    // will be maintained
+    final Deque<ScanResultValue> sortedElements = new ArrayDeque<>(q.size());
+    while (q.size() != 0) {
+      // We add at the front of the list because poll removes the tail of the queue.
+      sortedElements.addFirst(q.poll());
```

Comment:
Just want to double check but ArrayDeque#addFirst() is O(1), right?  Initially used a LinkedList but Forbidden APIs said no.

Reply:
This is the implementation:

```java
    public void addFirst(E e) {
        if (e == null)
            throw new NullPointerException();
        elements[head = (head - 1) & (elements.length - 1)] = e;
        if (head == tail)
            doubleCapacity();
    }
```

It looks O(1) on an amortized basis to me. Most of the function is O(1), except for `doubleCapacity()`, which is O(n) but it runs at most once every _n_ additions.